apps/framework-cli/src/infrastructure/redis/redis_client.rs

```diff
@@ -61,768 +176,788 @@ impl Default for RedisConfig {
-        let broadcast_channel = self.service_prefix(&["msgchannel"]);
+        let presence_task = tokio::spawn(async move {
+            let mut interval = tokio::time::interval(tokio::time::Duration::from_secs(1));
+            loop {
+                interval.tick().await;
+                if let Err(e) = presence_manager
+                    .update_presence(connection_manager.connection.clone())
+                    .await
+                {
+                    log::error!("<NewRedisClient> Error updating presence: {}", e);
+                }
+            }
+        });
 
-        info!(
-            "<RedisClient> Listening for messages on channels: {} and {}",
-            instance_channel, broadcast_channel
-        );
+        // Store the presence task
+        let presence_task_clone = self.presence_task.clone();
+        tokio::spawn(async move {
+            let mut presence_task_guard = presence_task_clone.write().await;
+            *presence_task_guard = Some(presence_task);
+        });
 
-        // Create a separate PubSub connection for listening to messages
-        let client = match Client::open(self.redis_config.url.clone()) {
-            Ok(client) => client,
-            Err(e) => {
-                error!("<RedisClient> Failed to open Redis client: {}", e);
-                return Err(e);
-            }
-        };
+        // Start connection monitor task
+        self.start_connection_monitor();
+    }
 
-        let pubsub_conn = match client.get_async_connection().await {
-            Ok(conn) => conn,
-            Err(e) => {
-                error!(
-                    "<RedisClient> Failed to get async connection for PubSub: {}",
-                    e
-                );
-                return Err(e);
-            }
-        };
+    pub fn start_connection_monitor(&self) -> tokio::task::JoinHandle<()> {
+        // Create a closure that builds the monitoring task
+        let mut connection_manager = self.connection_manager.clone();
+        let config = self.config.clone();
 
-        let mut pubsub = pubsub_conn.into_pubsub();
+        tokio::spawn(async move {
+            loop {
+                tokio::time::sleep(Duration::from_secs(5)).await;
+                if !connection_manager.ping().await {
+                    log::warn!("Redis connection lost, attempting reconnection");
+                    connection_manager.attempt_reconnection(&config).await;
+                }
+            }
+        })
+    }
 
-        pubsub
-            .subscribe(&[&instance_channel, &broadcast_channel])
-            .await
-            .map_err(|e| {
-                error!("<RedisClient> Failed to subscribe to channels: {}", e);
-                e
-            })?;
+    pub fn stop_periodic_tasks(&mut self) -> anyhow::Result<()> {
+        let rt = tokio::runtime::Handle::try_current()?;
 
-        let callback = self.message_callbacks.clone();
+        rt.block_on(async {
```

Comment:
Why do this need to be blocking?

Reply:
It's currently safe to remove this block.  Refactor will be part of next commit.